[wasm] Switch default modules to es6

pavelsavara opened the pull request:
## Motivation, benefits 
- TL;DR: 
    - consistent in NodeJS and in browsers
    - module isolation
    - well defined exports/imports
    - dependencies resolution
    - dependencies detected during parsing, not during run
    - async dynamic imports awaitable
- [Why Modules](https://eloquentjavascript.net/10_modules.html)
- Supported everywhere for many years, everywhere where WASM is able to run.

## Challenges, downsides
- CommonJS module could be [wrapped inside simple ES6 script](https://redfin.engineering/node-modules-at-war-why-commonjs-and-es-modules-cant-get-along-9617135eeca1). **But not the other way around.**
    - ES6 modules could be re-packaged by standard tools like webpack, rollup back to CJS or UMD.
    - libraries using another ES6 modules have to be also ES6 module (unless they re-pack with tools above).
- Node.js 12 doesn't support it, but is end-of-life by 2022-04-30
- After this PR will would still have way how to produce CommonJS, but [is it worth supporting it](https://github.com/dotnet/runtime/issues/63337) ?

## Changes in this PR
- switched dotnet.js to be ES6 module by making `<WasmEnableES6>` default true
- updated all samples
- updated functional tests
- updated debugger tests
- updated test-main to
- `EXPORT_ES6=1` should not be in the `emcc-link.rsp` because it could not be removed by workload re-link when `<WasmEnableES6> == false`.

pavelsavara commented:
/azp run runtime-manual

pavelsavara commented:
/azp run runtime-manual

pavelsavara committed:
- switched dotnet.js to be ES6 module as default
- updated all samples
- updated functional tests
- updated debugger tests

pavelsavara commented:
/azp run runtime-manual

pavelsavara committed:
fix

pavelsavara commented:
/azp run runtime-manual

pavelsavara committed:
fix

pavelsavara committed:
no message

pavelsavara commented:
/azp run runtime-manual

kg commented on src/mono/sample/mbr/DeltaHelper/DeltaHelper.csproj:
```diff
@@ -1,5 +1,5 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
+    <TargetFramework>net6.0</TargetFramework>
```
why? considering we're moving towards 7 this seems odd

pavelsavara replied:
Thanks, that was a mistake.

radical commented:
I can review once this is out of draft.

pavelsavara committed:
EXPORT_ES6=1 should not be in the emcc-default.rsp

pavelsavara committed:
feedback

pavelsavara commented:
/azp run runtime-manual

pavelsavara committed:
fix

pavelsavara commented:
/azp run runtime-manual

maraf committed:
Fix Wasm.Console.Node.CJS.Sample.

maraf committed:
Fix typescript sample on node.

- Copy typescript definition.
- Copy package.json from runtime.
- Change WasmMainJSPath.

maraf committed:
Fix HTML heading on WebPack sample.

maraf committed:
Add extra comment on typescript node sample about runtime location.

maraf commented:
WebPack sample doesn't work right now.

The reason is that webpack replaces `import.meta.url` with a compile time constant (= path on disk, eg. `file:///C:/...`). The runtime is than trying to load all assets as local files. This webpack behavior is an open issue https://github.com/webpack/webpack/issues/14445.

maraf commented:
/azp run runtime-manual

maraf committed:
Fix relink when ES6 is enabled.

maraf commented:
/azp run runtime-extra-platforms

maraf committed:
Provide a default locateFile which works in webpack.

Disable "module" compile time import checker.

maraf committed:
Small tweaks in webpack sample.

maraf commented:
/azp run runtime-extra-platforms

maraf committed:
Revert "Provide a default locateFile which works in webpack."

This reverts commit 25ee613f177ba895c9326120480a0ab43dd868df.

pavelsavara committed:
Merge branch 'main' into wasm_switch_to_es6

pavelsavara committed:
templates, webpack

pavelsavara committed:
fix webpack, add webpack nodejs

pavelsavara committed:
Merge branch 'main' into wasm_switch_to_es6

# Conflicts:
#	src/mono/wasm/runtime/CMakeLists.txt
#	src/mono/wasm/wasm.proj

pavelsavara commented:
/azp run runtime-extra-platforms

pavelsavara committed:
Merge branch 'main' into wasm_switch_to_es6

# Conflicts:
#	src/mono/wasm/test-main.js

pavelsavara commented:
/azp run runtime-extra-platforms

pavelsavara committed:
fix

pavelsavara commented:
/azp run runtime-extra-platforms

pavelsavara committed:
fix

pavelsavara committed:
Merge branch 'main' into wasm_switch_to_es6

pavelsavara committed:
fix
